akka-distributed-data/src/main/scala/akka/cluster/ddata/ORSet.scala

```diff
@@ -289,24 +368,70 @@ final class ORSet[A] private[akka] (
   override def merge(that: ORSet[A]): ORSet[A] = {
     if ((this eq that) || that.isAncestorOf(this)) this.clearAncestor()
     else if (this.isAncestorOf(that)) that.clearAncestor()
-    else {
-      val commonKeys =
-        if (this.elementsMap.size < that.elementsMap.size)
-          this.elementsMap.keysIterator.filter(that.elementsMap.contains)
-        else
-          that.elementsMap.keysIterator.filter(this.elementsMap.contains)
-      val entries00 = ORSet.mergeCommonKeys(commonKeys, this, that)
-      val thisUniqueKeys = this.elementsMap.keysIterator.filterNot(that.elementsMap.contains)
-      val entries0 = ORSet.mergeDisjointKeys(thisUniqueKeys, this.elementsMap, that.vvector, entries00)
-      val thatUniqueKeys = that.elementsMap.keysIterator.filterNot(this.elementsMap.contains)
-      val entries = ORSet.mergeDisjointKeys(thatUniqueKeys, that.elementsMap, this.vvector, entries0)
-      val mergedVvector = this.vvector.merge(that.vvector)
-
-      clearAncestor()
-      new ORSet(entries, mergedVvector)
+    else dryMerge(that, addDeltaOp = false)
+  }
+
+  // share merge impl between full state merge and AddDeltaOp merge
+  private def dryMerge(that: ORSet[A], addDeltaOp: Boolean): ORSet[A] = {
+    val commonKeys =
+      if (this.elementsMap.size < that.elementsMap.size)
+        this.elementsMap.keysIterator.filter(that.elementsMap.contains)
+      else
+        that.elementsMap.keysIterator.filter(this.elementsMap.contains)
+    val entries00 = ORSet.mergeCommonKeys(commonKeys, this, that)
+    val entries0 =
+      if (addDeltaOp)
+        entries00 ++ this.elementsMap.filter { case (elem, _) ⇒ !that.elementsMap.contains(elem) }
+      else {
+        val thisUniqueKeys = this.elementsMap.keysIterator.filterNot(that.elementsMap.contains)
+        ORSet.mergeDisjointKeys(thisUniqueKeys, this.elementsMap, that.vvector, entries00)
+      }
+    val thatUniqueKeys = that.elementsMap.keysIterator.filterNot(this.elementsMap.contains)
+    val entries = ORSet.mergeDisjointKeys(thatUniqueKeys, that.elementsMap, this.vvector, entries0)
+    val mergedVvector = this.vvector.merge(that.vvector)
+
+    clearAncestor()
+    new ORSet(entries, mergedVvector)
+  }
+
+  override def mergeDelta(thatDelta: ORSet.DeltaOp): ORSet[A] = {
+    thatDelta match {
+      case d: ORSet.AddDeltaOp[A]    ⇒ dryMerge(d.underlying, addDeltaOp = true)
+      case d: ORSet.RemoveDeltaOp[A] ⇒ mergeRemoveDelta(d)
+      case ORSet.DeltaGroup(ops) ⇒
+        ops.foldLeft(this) {
+          case (acc, op: ORSet.AddDeltaOp[A])    ⇒ acc.dryMerge(op.underlying, addDeltaOp = true)
+          case (acc, op: ORSet.RemoveDeltaOp[A]) ⇒ acc.mergeRemoveDelta(op)
+          case (acc, op: ORSet.DeltaGroup[A]) ⇒
+            throw new IllegalArgumentException("ORSet.DeltaGroup should not be nested")
+        }
     }
   }
 
+  private def mergeRemoveDelta(thatDelta: ORSet.RemoveDeltaOp[A]): ORSet[A] = {
```

Comment:
@gosubpl This is the part I'm most uncertain about. I found a bug in previous implementation via the random test `ReplicatorDeltaSpec` and could reproduce it in unit test https://github.com/akka/akka/pull/22331/files#diff-d020812bcac00bfa1009224c2315c362R334